File: Arrays & Hashing/top_k_frequent_elements.py
"""
https://leetcode.com/problems/top-k-frequent-elements/description/
"""

class MySolution: #Time complexity O(klogn), Space complexity O(n)
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        arr = []
        counter = {}

        for num in nums:
            counter[num] = counter.get(num, 0) - 1

        arr = list(zip(counter.values(), counter.keys()))

        heapq.heapify(arr)

        res = []

        while k:
            freq, value = heapq.heappop(arr)
            res.append(value)
            k -= 1

        return res
    

class OptimalSolution: #Time complexity O(n), Space complexity O(n)
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counter = {}
        freq = [[] for i in range(len(nums) + 1)] #we use the index position as a reference for the freqency count - [] = values that have a freq of the index position

        for num in nums:
            counter[num] = counter.get(num, 0) + 1

        for key, value in counter.items():
            freq[value].append(key)

        res = []

        for i in range(len(freq) - 1, 0, -1):
            while freq[i] and k:
                value = freq[i].pop()
                res.append(value)
                k -= 1

            if k == 0:
                break

        return res

            
class MyOtherSolution: #Same complexity as above
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counts = defaultdict(int)

        for num in nums:
            counts[num] += 1


        freqs = [[] for _ in range(len(nums) + 1)]

        for num, freq in counts.items():

            freqs[freq].append(num)

        res = []

        for i in range(len(freqs) - 1, -1, -1):
            if freqs[i]:
                for num in freqs[i]:
                    res.append(num)
                    if len(res) == k:
                        return res

         
        

            


         
        

            

File: Arrays & Hashing/top_k_frequent_elements.js
/*
https://leetcode.com/problems/top-k-frequent-elements/description/
*/

class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number[]}
     * Time complexity O(n), Space complexity O(n)
     */
    topKFrequent(nums, k) {
        const counts = new Map()

        for(var num of nums){
            if(!counts.has(num)){
                counts.set(num, 0)
            }

            counts.set(num, counts.get(num) + 1)
        }

        const res = []

        const freqs = new Array(nums.length + 1).fill(null).map(() => [])

        for (let [num, count] of counts) {
            freqs[count].push(num)
        }

        for(let i = freqs.length - 1; i > -1; i--){
            if(freqs[i]){
                for(var num of freqs[i]){
                    res.push(num)
                    if (res.length === k){
                        return res
                    }
                }
            }
        }


    }
}
